File: backend/SmartSolarMicrogrid.Api/Services/Dashboard/DashboardService.cs
/*
 * File Name: DashboardService.cs
 * Project: Smart Solar Microgrid Trading System
 * Module: SE4040 Enterprise Application Development
 * Author: IT22154880
 * Description: Implementation of DashboardService.cs
 * Date: 2026-09-19
 */

using SmartSolarMicrogrid.Api.Models.DTOs;
using SmartSolarMicrogrid.Api.Models.Enums;
using SmartSolarMicrogrid.Api.Repositories;

namespace SmartSolarMicrogrid.Api.Services.Dashboard;

public class DashboardService : IDashboardService
{
    private readonly IEnergyReservationRepository _reservationRepository;

    public DashboardService(IEnergyReservationRepository reservationRepository)
    {
        _reservationRepository = reservationRepository;
    }

    public async Task<DashboardStatsDto> GetStatsAsync(string role, string? userNic)
    {
        // Retrieves stats data from the system.
        var allReservations = await _reservationRepository.GetAllAsync();
        
        var query = allReservations.AsEnumerable();

        if (role == "PROSUMER" && !string.IsNullOrEmpty(userNic))
        {
            query = query.Where(r => r.ProsumerNic == userNic);
        }

        var pendingCount = query.Count(r => r.Status == ReservationStatus.PENDING);
        var approvedFutureCount = query.Count(r => r.Status == ReservationStatus.APPROVED && r.ScheduledStartDateTime.Date >= DateTime.UtcNow.Date);

        return new DashboardStatsDto
        {
            PendingReservationsCount = pendingCount,
            ApprovedFutureReservationsCount = approvedFutureCount
        };
    }
}

File: web/SmartSolarMicrogrid.Web/src/services/dashboardService.js
/*
 * File Name: dashboardService.js
 * Project: Smart Solar Microgrid Trading System
 * Module: SE4040 Enterprise Application Development
 * Author: Isara
 * Description: API service for fetching dashboard metrics and data.
 * Date: 2026-09-14
 */

import api from './api';

const dashboardService = {
  // Fetch all stations
  getStations: async () => {
    try {
      const response = await api.get('/Stations');
      return response.data;
    } catch (error) {
      console.error('Error fetching stations:', error);
      return { success: false, data: [] };
    }
  },

  // Fetch all users
  getUsers: async () => {
    try {
      const response = await api.get('/Users');
      return response.data;
    } catch (error) {
      console.error('Error fetching users:', error);
      return { success: false, data: [] };
    }
  },

  // Fetch all prosumers (could also just use users and filter)
  getProsumers: async () => {
    try {
      const response = await api.get('/Prosumers');
      return response.data;
    } catch (error) {
      console.error('Error fetching prosumers:', error);
      return { success: false, data: [] };
    }
  },

  // Fetch pending prosumer activations
  getPendingActivations: async () => {
    try {
      const response = await api.get('/Prosumers/pending');
      return response.data;
    } catch (error) {
      console.error('Error fetching pending activations:', error);
      return { success: false, data: [] };
    }
  },

  // Fetch all reservations
  getReservations: async () => {
    try {
      const response = await api.get('/Reservations');
      return response.data;
    } catch (error) {
      console.error('Error fetching reservations:', error);
      return { success: false, data: [] };
    }
  }
};

export default dashboardService;
